### PR TITLE
added Adagrad optimizer ...

### LINKED ISSUE #46: Extend mlx with Additional Optimizers
**Issue Type:** Enhancement

**Proposal:**
I would like to propose an enhancement to the framework by adding a variety of optimizers. While the current implementation includes Stochastic Gradient Descent (SGD) and Adam optimizers, the addition of more optimization algorithms could greatly benefit users with diverse requirements.

**Potential Optimizers to Include:**
1. **Adagrad:** 
2. **RMSprop:** 
3. **Adadelta:** 
4. **Nadam:** 
5. **AdamW**
etc ...

I would appreciate your guidance on the next steps for contributing to this enhancement. Whether it's providing additional details, discussing the proposed optimizers, or collaborating on the implementation, I am eager to contribute to the growth of this fantastic framework.

### PR DESCRIPTION
**Pull request by m0saan**
**Pull Request Description:**

This pull request adds support for the Adagrad optimizer to the mlx set of optimizers, aligning with the implementation convention used for existing optimizers in the repository.

Fixes #46

### CODE REVIEW

File: python/mlx/optimizers.py [hunk 1]
```diff
@@ -221,3 +221,44 @@ def apply_single(
         return super().apply_single(
             gradient, parameter * (1 - self.learning_rate * self.weight_decay), state
         )
+
+
+class Adagrad(Optimizer):
+    r"""Implementation of the Adagrad optimizer [1].
+
+    Our Adagrad implementation follows the original paper and omits the bias
+    correction in the first and second moment estimates. In detail,
```
Comment (angeloskath) [inline python/mlx/optimizers.py:230]: I think this has been left from copying from Adam as there are no moment estimates.

File: python/mlx/optimizers.py [hunk 2]
```diff
@@ -221,3 +221,44 @@ def apply_single(
...
+
+    def apply_single(
+        self, gradient: mx.array, parameter: mx.array, state: OptimizerState
+    ):
+        """Performs the Adagrad parameter update and stores :math:`v` in the
+        optimizer state."""
+        lr = self.learning_rate
+        eps = self.eps
+
+        v = state.get("v", mx.square(gradient))
```
Comment (angeloskath) [inline python/mlx/optimizers.py:260]: I think this should be `zeros_like` instead of square gradient.

File: python/mlx/optimizers.py [hunk 3]
```diff
@@ -221,3 +221,44 @@ def apply_single(
...
+    def __init__(self, learning_rate: float, eps: float = 1e-8):
+        super().__init__()
+
+        self.learning_rate = learning_rate
+        self.eps = eps
+
+        if self.learning_rate < 0.0:
+            raise ValueError("Invalid learning rate: {}".format(self.learning_rate))
+        if self.eps < 0.0:
+            raise ValueError("Invalid eps: {}".format(self.eps))
```
Comment (angeloskath) [inline python/mlx/optimizers.py:250]: Maybe too nitpicky but we use format strings almost everywhere in the library for things like that. I think it is slightly more readable as well. For instance,

```python
raise ValueError(f"Adagrad epsilon should be >0, {self.eps} was provided instead")
```

**Review by angeloskath — CHANGES_REQUESTED** [review]
Thanks, that looks great.

I left some minor comments.

File: python/mlx/optimizers.py (hunk 3 above)
Comment (m0saan) [reply python/mlx/optimizers.py:249]: done!

File: python/mlx/optimizers.py (hunk 2 above)
Comment (m0saan) [reply python/mlx/optimizers.py:259]: done!

**Comment by m0saan** [comment]
> Thanks, that looks great.
> 
> I left some minor comments.

Thanks so much for the comments! :p

**Review by angeloskath — APPROVED** [review]
